README.md
# django-payments-mercadopago

A mercadopago payment gateway backend for [django-payments](https://github.com/mirumee/django-payments)

## Installation

You can install it via pip

```Shell
pip install django-payments-mercadopago
```

Add *payments_mercadopago* to your *settings.py* file

```python
INSTALLED_APPS = [
    # ...
    'payments_mercadopago',
    ]
```

## Settings.py configuration

Add the *payments_mercadopago.MercadoPagoProvider* to your *PAYMENT_VARIANTS* variable. Also to make it available add MercadoPago to your *CHECKOUT_PAYMENT_CHOICES variable*

Mercado Pago doesn't require you to specify a fixed sandbox endpoint, instead they provide you with a sandbox token and a production token, for testing and production, respectively.

### Configuration for development

```Python
PAYMENT_VARIANTS = {
    # ...
    'MercadoPago':('payments_mercadopago.MercadoPagoProvider',{
        'access_token': 'MERCADO_PAGO_SANDBOX_ACCESS_TOKEN',
        'sandbox_mode': True})
}

CHECKOUT_PAYMENT_CHOICES = [('MercadoPago', 'Mercado Pago')]
```

If you have any problem using localhost urls as the return value of get_failure_url() or get_success_url() methods try using [ngrok](https://ngrok.com/) instead.

### Configuration for production

```Python
PAYMENT_VARIANTS = {
    # ...
    'MercadoPago':('payments_mercadopago.MercadoPagoProvider',{
        'access_token': 'MERCADO_PAGO_ACCESS_TOKEN',
        'sandbox_mode': False})
}

CHECKOUT_PAYMENT_CHOICES = [('MercadoPago', 'Mercado Pago')]
```

## Obtaining the Tokens

You can get your own Mercado Pago production and sandbox access tokens in your [Mercado Pago developer panel](https://www.mercadopago.com.mx/developers/panel/credentials)


## Documentation

For detailed instructions on how to use django-payments please visit the official [django-payments documentation](https://django-payments.readthedocs.io/en/latest/)



HISTORY.rst
=======
History
=======

0.4.0 (2020-10-17)
------------------

* Change project layout, adapted from cookiecutter. Include tox tests, add AUTHOR, CONTRIBUTING, HISTORY, Docs, etc.


test_settings.py
import os

PROJECT_ROOT = os.path.normpath(
    os.path.join(os.path.dirname(__file__), 'payments_mercadopago'))
TEMPLATES = [{
    'BACKEND': 'django.template.backends.django.DjangoTemplates',
    'DIRS': [os.path.join(PROJECT_ROOT, 'templates')]}]

SECRET_KEY = 'MY-SECRET-KEY'
PAYMENT_HOST = 'example.org'

INSTALLED_APPS = ['payments', 'django.contrib.sites']
ROOT_URLCONF = 'payments.urls'


payments_mercadopago/test_mercadopago.py
from __future__ import unicode_literals
import json
from decimal import Decimal
from unittest import TestCase
from mock import patch, MagicMock, Mock

from django.http import HttpResponse
from django.utils import timezone

from . import MercadoPagoProvider
from payments import PurchasedItem, RedirectNeeded, PaymentError, PaymentStatus

CLIENT_ID = 'Mercado Pago Test User'
PAYMENT_TOKEN = '5a4dae68-2715-4b1e-8bb2-2c2dbe9255f6'
ACCESS_TOKEN = 'TEST_123456789ABCDEFGHIJKLM'
INIT_POINT = 'sandbox_init_point'
SANDBOX_INIT_POINT_URL = 'http://sandbox.mercadopago.com/redirecttopayment'
INIT_POINT_URL = 'http://mercadopago.com/redirecttopayment'
VARIANT = 'wallet'
CURRENCY = 'MXN'

class Payment(Mock):
    id = 1
    description = 'payment'
    currency = CURRENCY
    delivery = Decimal(10)
    status = PaymentStatus.WAITING
    tax = Decimal(10)
    token = PAYMENT_TOKEN
    total = Decimal(100)
    captured_amount = Decimal(0)
    variant = VARIANT
    transaction_id = None
    message = ''
    extra_data = {}
    billing_first_name = 'John'
    billing_last_name = 'Doe'
    billing_email = 'false@email.com'
    billing_address_1 = 'false st'
    billing_address_2 = '111'
    billing_postcode = '00000'

    def change_status(self, status, message=''):
        self.status = status
        self.message = message

    def get_failure_url(self):
        return 'http://cancel.com'

    def get_process_url(self):
        return 'http://example.com'

    def get_purchased_items(self):
        return [
            PurchasedItem(
                name='foo', quantity=Decimal('10'), price=Decimal('20'),
                currency='MXN', sku='bar')]

    def get_success_url(self):
        return 'http://success.com'

class TestMercadoPagoProvider(TestCase):

    def setUp(self):
        self.payment = Payment()
        self.provider = MercadoPagoProvider(
            access_token=ACCESS_TOKEN, sandbox_mode=True)

    @patch('mercadopago.MP.create_preference')
    def test_provider_raises_redirect_needed_on_success_preference_creation(
            self, mocked_create_preference):
        mocked_create_preference.return_value = {
            'status': 201,
            'response': {
                'sandbox_init_point': SANDBOX_INIT_POINT_URL,
                'init_point': INIT_POINT_URL
            }
        }
        with self.assertRaises(RedirectNeeded) as exc:
            self.provider.get_form(
                payment=self.payment)
            self.assertEqual(exc.args[0], SANDBOX_INIT_POINT_URL)
        self.assertEqual(self.payment.status, PaymentStatus.WAITING)
        self.assertEqual(self.payment.captured_amount, Decimal('0'))

    def test_process_data_returns_200_on_http_post_request(self):
        request = MagicMock()
        request.GET = {}
        response = self.provider.process_data(self.payment, request)
        self.assertTrue(isinstance(response, HttpResponse))
        self.assertEqual(response.status_code, 200)
        self.assertEqual(self.payment.captured_amount, Decimal('0'))
        self.assertEqual(self.payment.status, PaymentStatus.WAITING)

    @patch('mercadopago.MP.get_payment')
    def test_process_data_handles_payment_info(self, mocked_get_payment):
        request = MagicMock()
        request.GET = {'data.id': '123456', 'type': 'payment'}
        mocked_get_payment.return_value = {
            'status': 200,
            'response': {
                'status': 'approved'
            }
        }
        response = self.provider.process_data(self.payment, request)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(self.payment.captured_amount, self.payment.total)
        self.assertEqual(self.payment.status, PaymentStatus.CONFIRMED)
        self.assertEqual(self.payment.transaction_id, '123456')
        self.assertEqual(self.payment.extra_data,
            json.dumps(mocked_get_payment.return_value))

    @patch('mercadopago.MP.get_payment')
    def test_process_data_handles_no_payment_post_requests(
        self, mocked_get_payment):
        request = MagicMock()
        request.GET = {'another_post_request': 'no_data',
            'another_data': 'other_data'}
        mocked_get_payment.return_value = {
            'status': 200,
            'response': {
                'status': 'approved'
            }
        }
        response = self.provider.process_data(self.payment, request)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(self.payment.captured_amount, Decimal('0'))
        self.assertEqual(self.payment.status, PaymentStatus.WAITING)

    @patch('mercadopago.MP.refund_payment')
    def test_provider_refunds_payment(self, mocked_get_refund_payment):
        mocked_get_refund_payment.return_value ={
            'status': 201,
            'response': {
                'status': 'approved'
            }
        }
        self.provider.refund(self.payment)
        self.assertEqual(self.payment.captured_amount, Decimal('0'))
        self.assertEqual(self.payment.status, PaymentStatus.REFUNDED)
        self.assertEqual(self.payment.extra_data,
            json.dumps(mocked_get_refund_payment.return_value))

    @patch('mercadopago.MP.create_preference')
    def test_create_payment_return_preference(self, mocked_create_preference):
        mocked_create_preference.return_value = {
            'status': 200,
            'response':{
                'items': ['items'],
                'notification_url': 'https://yourhost.com/process_response',
                'back_urls': {
                    'success': 'https://success.com',
                    'failure': 'https://failure.com',
                    'pending': 'https://pending.com'
                }
            }
        }
        created_payment = self.provider.create_payment(self.payment)
        self.assertEqual(created_payment, mocked_create_preference.return_value)
        self.assertEqual(self.payment.extra_data,
            json.dumps(mocked_create_preference.return_value))

    def test_create_preference_data_returns_dictionary(self):
        result = self.provider.create_preference_data(self.payment)
        self.assertTrue(isinstance(result, dict))
        self.assertEqual(result['back_urls']['success'],
            self.payment.get_success_url())
        self.assertEqual(result['back_urls']['failure'],
            self.payment.get_failure_url())
        self.assertEqual(result['external_reference'], self.payment.token)

    @patch('mercadopago.MP.get_payment')
    def test_get_payment_information_return_payment_details(self,
        mocked_get_payment):
        mocked_get_payment.return_value = {
            'status': 200,
            'response': {
            }
        }
        result = self.provider.get_payment_information(
            self.payment.transaction_id)
        self.assertTrue(isinstance(result, dict))
        self.assertEqual(result, mocked_get_payment.return_value)

    @patch('mercadopago.MP.get_payment')
    def test_get_payment_raises_paymentError_on_404(self, mocked_get_payment):
        mocked_get_payment.return_value = {
            'status': 404,
            'response': {
                'message': 'Payment not found',
                'error': 'not found'
            }
        }
        with self.assertRaises(PaymentError) as exc:
            self.provider.get_payment_information(self.payment.transaction_id)

    @patch('mercadopago.MP.create_preference')
    def test_create_payment_raises_paymentError_on_400(
        self, mocked_create_preference):
        mocked_create_preference.return_value = {
            'status': 400,
            'response': {
                'message': 'items needed',
                'error': 'invalid items'
            }
        }
        with self.assertRaises(PaymentError) as exc:
            self.provider.create_payment(self.payment)


payments_mercadopago/__init__.py
from django.urls import reverse
from django.http import HttpResponse, HttpResponseForbidden
from django.shortcuts import redirect
from django.utils.translation import gettext as _

from decimal import Decimal, ROUND_HALF_UP
import json
import logging
try:
    from urllib.parse import urljoin
except ImportError:
    from urlparse import urljoin

from payments import PaymentError, PaymentStatus, RedirectNeeded
from payments.core import BasicProvider, get_base_url

import mercadopago

CENTS = Decimal('0.01')

logger = logging.getLogger(__name__)


class MercadoPagoProvider(BasicProvider):

    def __init__(self, access_token, sandbox_mode=False, **kwargs):
        self.access_token = access_token
        self.sandbox_mode = sandbox_mode
        self.init_point = 'sandbox_init_point' if self.sandbox_mode else 'init_point'
        self.mp = mercadopago.MP(self.access_token)
        self.mp.sandbox_mode(self.sandbox_mode)
        super(MercadoPagoProvider, self).__init__(**kwargs)

    def get_form(self, payment, data=None, file_data=None):
        if not payment.id:
            payment.save()
        payment_data = self.create_payment(payment)
        redirect_to = self.get_value_from_response(
            payment_data, self.init_point)
        payment.change_status(PaymentStatus.WAITING)
        raise RedirectNeeded(redirect_to)

    def create_payment(self, payment):
        preference = self.create_preference_data(payment)
        preferenceResult = self.mp.create_preference(preference)
        payment.extra_data = json.dumps(preferenceResult)
        if 200 <= preferenceResult['status'] <= 201:
            return preferenceResult
        else:
            message = self.get_value_from_response(
                preferenceResult, 'message')
            logger.warning(message, extra={
                           "response": preferenceResult})
            raise PaymentError(message)

    def get_value_from_response(self, response, key):
        return response.get('response', {}).get(key, {})

    def create_notification_url(self, payment):
        return urljoin(get_base_url(), reverse('process_payment',
                                               kwargs={"token": payment.token}))

    def create_preference_data(self, payment):
        items = list(self.get_transactions_items(payment))
        items.insert(
            0, self.get_order_name_and_shipping_cost(payment))
        sub_total = (
            payment.total - payment.delivery - payment.tax)
        sub_total = sub_total.quantize(CENTS, rounding=ROUND_HALF_UP)
        total = payment.total.quantize(CENTS, rounding=ROUND_HALF_UP)
        tax = payment.tax.quantize(CENTS, rounding=ROUND_HALF_UP)
        delivery = payment.delivery.quantize(
            CENTS, rounding=ROUND_HALF_UP)
        full_name = payment.billing_first_name + " " \
            + payment.billing_last_name
        full_address = payment.billing_address_1 + " " \
            + payment.billing_address_2
        # Please check the preferences API for customization
        # www.mercadopago.com.mx/developers/es/reference/preferences/resource/
        preferenceData = {
            "items": items,
            "payer": {
                "name": full_name,
                "email": payment.billing_email,
                "address": {
                    "street_name": full_address,
                    "zip_code": payment.billing_postcode
                },
            },

            "back_urls": {
                # Localhost urls raise an error in payment in
                # mercadopago website, even in sandbox mode
                # When testing in localhost use Ngrok instead
                "success": payment.get_success_url(),
                "failure": payment.get_failure_url(),
                "pending": payment.get_failure_url(),
            },
            "auto_return": "approved",
            "notification_url": self.create_notification_url(payment),
            "external_reference": payment.token,
        }
        return preferenceData

    def get_transactions_items(self, payment):
        for purchased_item in payment.get_purchased_items():
            price = purchased_item.price.quantize(
                CENTS, rounding=ROUND_HALF_UP)
            item = {'title': purchased_item.name[:127],
                    'quantity': purchased_item.quantity,
                    'unit_price': float(price),
                    'currency_id': purchased_item.currency,
                    'id': purchased_item.sku}
            yield item

    def get_order_name_and_shipping_cost(self, payment):
        item = {'title': payment.description + _(' and shipping'),
                'quantity': 1,
                'unit_price': float(payment.delivery),
                'currency': payment.currency,
                'id': _('shipping')
                }
        return item

    def process_data(self, payment, request):
        if 'data.id' in request.GET and 'type' in request.GET:
            if request.GET.get('type') == 'payment':
                collection_id = request.GET.get('data.id')
                payment.transaction_id = collection_id
                payment_information = self.get_payment_information(
                    collection_id)
                payment.extra_data = json.dumps(payment_information)
                collection_status = self.get_value_from_response(
                    payment_information, "status")
                if collection_status == 'approved':
                    payment.captured_amount = payment.total
                    payment.change_status(PaymentStatus.CONFIRMED)
                elif collection_status == 'in_process':
                    payment.change_status(PaymentStatus.WAITING)
        return HttpResponse(status=200)

    def get_payment_information(self, payment_id):
        paymentInfo = self.mp.get_payment(payment_id)
        if paymentInfo["status"] == 200:
            return paymentInfo
        else:
            message = self.get_value_from_response(
                paymentInfo, 'message')
            logger.warning(message, extra={"response": paymentInfo})
            raise PaymentError(message)

    def refund(self, payment, amount=None):
        amount = payment.captured_amount
        # MercadoPago Official Python SDK doesn't support partial refunds
        refundResult = self.mp.refund_payment(payment.transaction_id)
        payment.extra_data = json.dumps(refundResult)
        if refundResult['status'] == 201:
            payment.change_status(PaymentStatus.REFUNDED)
            return amount
        else:
            message = self.get_value_from_response(
                refundResult, 'message')
            raise PaymentError(message)

    def cancel(self, payment):
        cancelationResult = self.mp.cancel_payment(
            payment.transaction_id)
        payment.extra_data = json.dumps(cancelationResult)
        if cancelationResult['status'] == 200:
            payment.change_status(PaymentStatus.REJECTED)
        else:
            message = self.get_value_from_response(
                cancelationResult, 'message')
            logger.warning(message, extra={
                           "response": cancelationResult})
            raise PaymentError(message)
